File: Code/src/portfoliomanager/model/Holding.java
package portfoliomanager.model;

import java.math.BigDecimal;
import java.math.RoundingMode;

/**
 * Crypto Holdings 
 * @author Colby
 * @version Spring 2025
 */
public class Holding extends Crypto {
	private double amount;
	/**
	 * Creates an instance of a Holding
	 * @param name the name 
	 * @param currentPrice the current price
	 * @param amount the amount held
	 */
	
	public Holding(String name, Double currentPrice, double amount) {
		super(name, currentPrice);
		this.amount = amount;
	}
	
	/**
	 * Gets the total price 
	 * @return the price of the asset
	 */
	public double getTotalPrice() {
		return super.getCurrentPrice() * this.amount;
	}
	
	@Override
	public String toString() {
		return String.format("%10s%25s%25s", super.getName(), this.getTotalPrice(), this.amount);
	}
	/**
	 * Gets the amount held
	 * @return the amount held
	 */
	
	public double getAmountHeld() {
		return this.amount;
	}
	/**
	 * Sets the ammount held
	 * @param amount the amount to set
	 */
	
	public void setAmountHeld(double amount) {
		this.amount = this.roundToTwoDecimalPlaces(amount);

	}
	
	private double roundToTwoDecimalPlaces(double value) {
		BigDecimal bd = new BigDecimal(value);
		bd = bd.setScale(2, RoundingMode.HALF_UP);
		return bd.doubleValue();
	}
	
	/**
	 * Gets the profit by selling a portion of it 
	 * @param amountToSell the amount you want to sell
	 * @return the total gain from selling
	 */
	
	public double getProfit(double amountToSell) {
		return super.getCurrentPrice() * amountToSell;
	}
}

File: Code/src/portfoliomanager/model/Holding.py
'''
Created on Apr 3, 2025

@author: Colby
'''
class Holding:
    '''
    Instantiates a holding
    @post self.amount == amount, self.name == name
    '''
    def __init__(self, name, amount):
        self.amount = amount
        self.name = name
        
    def set_amount_held(self, amount: float):
        '''
        Sets the amount held
        @post self.amount = amount
        '''
        self.amount = amount
        
    def getHoldingName(self):
        '''
        Gets the name of the holding
        @returns self.name
        '''
        return self.name
    
    def getAmountHeld(self):
        '''
        Gets the amount held
        @returns the amount held
        '''
        return self.amount
        
    def to_dict(self):
        '''
        transforms the contents of the holding to a dictionary
        @returns a dictionary formatted as follows
        '''
        return {"name": self.name, "amount": self.amount}
